vendors/st/boards/stm32l475_discovery/ports/common_io/iot_i2c.c

```diff
@@ -0,0 +1,744 @@
+                                                                               uint16_t DevAddress,
+                                                                               uint8_t * pData,
+                                                                               uint16_t Size,
+                                                                               uint32_t XferOptions ) );
+
+/*--------------------API Implementation---------------------*/
+
+IotI2CHandle_t iot_i2c_open( int32_t lI2CInstance )
+{
+    /* Exceeded maximum instance number. */
+    if( lI2CInstance >= _I2C_INSTANCE_NUM )
+    {
+        return NULL;
+    }
+
+    IotI2CDescriptor_t * pI2cDescriptor = &_i2cContexts[ lI2CInstance ];
+    I2C_HandleTypeDef * pHalHandle = &( pI2cDescriptor->xHandle );
+    HAL_StatusTypeDef halStatus = HAL_OK;
+
+    /* Not opened yet. */
+    if( _isFlagNotSet( pI2cDescriptor, _FLAG_OPENED ) )
+    {
+        halStatus = HAL_I2C_Init( pHalHandle );
+
+        if( halStatus == HAL_OK )
+        {
+            /* Configure Analog filter. */
+            halStatus = HAL_I2CEx_ConfigAnalogFilter( pHalHandle, I2C_ANALOGFILTER_ENABLE );
+        }
+
+        if( halStatus == HAL_OK )
+        {
+            pI2cDescriptor->xSemphr = xSemaphoreCreateBinary();
+
+            /* If semaphore creation fails, set returned handle to NULL. */
+            if( pI2cDescriptor->xSemphr == NULL )
+            {
+                pI2cDescriptor = NULL;
+            }
+        }
+        else
+        {
+            /* In failure case, set returned handle to NULL. */
+            pI2cDescriptor = NULL;
+        }
+    }
+    else
+    {
+        /* Return NULL handle if opened more than once. */
+        pI2cDescriptor = NULL;
+    }
+
+    if( pI2cDescriptor != NULL )
+    {
+        /* Set opened flag. */
+        pI2cDescriptor->flags |= _FLAG_OPENED;
+    }
+
+    return pI2cDescriptor;
+}
+
+void iot_i2c_set_callback( IotI2CHandle_t const pxI2CPeripheral,
+                           IotI2CCallback_t xCallback,
+                           void * pvUserContext )
+{
+    IotI2CDescriptor_t * pI2cDescriptor = pxI2CPeripheral;
+
+    /* Silently returns if input handle is invalid. */
+    if( ( pI2cDescriptor == NULL ) || ( ( pI2cDescriptor->flags & _FLAG_OPENED ) == 0 ) )
+    {
+        return;
+    }
+
+    pI2cDescriptor->xCallback = xCallback;
+    pI2cDescriptor->pvUserContext = pvUserContext;
+}
+
+int32_t iot_i2c_read_sync( IotI2CHandle_t const pxI2CPeripheral,
+                           uint8_t * const pucBuffer,
+                           size_t xBytes )
+{
+    IotI2CDescriptor_t * pI2cDescriptor = pxI2CPeripheral;
+    HAL_StatusTypeDef halStatus = HAL_OK;
+
+    if( ( pI2cDescriptor == NULL ) ||
+        ( _isFlagNotSet( pI2cDescriptor, _FLAG_OPENED ) ) ||
+        ( pucBuffer == NULL ) ||
+        ( xBytes == 0 ) )
+    {
+        return IOT_I2C_INVALID_VALUE;
```

Comment:
Are we trying to restrict ourselves to a single return for MISRA compliance?

Reply:
I didn't follow some MISRA rules. Good callout. Let me create a SIM to track this issue.

Reply:
Why not fix it now?

Reply:
I can fix the "single return". But leave other MISRA violation later.